### ISSUE TITLE
Differing function call requirements between languages with pyfunction

### ISSUE THREAD

**Issue by JoshuaKeegan3** [mityax/rustimport#29]
Calling a function can cause compilation errors in python while fixing them in rust or vice versa. This is unintuitive and takes away from how seamless pyfunction is. It also prevents the user from testing the code in rust forcing them to test in python.

 eg: The following code will run with cargo run however not when run with a python file

```rust
#[pyfunction]
fn test() {
   let window = WindowDesc::new(build_app())
}```

While this code will run with a python file but not with cargo

```rust
#[pyfunction]
fn test() {
   let window = WindowDesc::new(build_app)
}
```

The difference is weather or not the function is called.

Something that is interesting to note is rust loses deep type inference in this situation with this set up

<img width="748" alt="Screenshot 2024-02-19 at 2 14 59 AM" src="https://github.com/mityax/rustimport/assets/37237625/ad049237-6b67-405d-9eb2-6494cc3cec99">

<img width="1020" alt="Screenshot 2024-02-19 at 2 15 59 AM" src="https://github.com/mityax/rustimport/assets/37237625/6ff9493f-0145-4d68-b9bf-9312a6c0f35d">

The final thing I found that is probably a different issue was some imports just disappeared when using the python version. Eg

` use druid::widget::AspectRatioBox` can't be imported into python saying that it doesn't exist however it can be used in the Rust version. This is a very specific error as it doesn't happen with other `rust druid::widget::foos`. It does however also happen with `druid::theme::TextColor`

<img width="774" alt="Screenshot 2024-02-19 at 2 28 03 AM" src="https://github.com/mityax/rustimport/assets/37237625/a9207be4-73b3-4ce8-b600-30d6a9800256">

**Comment by mityax** [comment]
Hi @JoshuaKeegan3, your issue should probably rather be directed at https://github.com/PyO3/pyo3, which is responsible for `pyfunction` and the rust-python bindings in general.

There is a (very small) chance that the import error you're mentioning has something to do with rustimport, but to asses that I need more context. Are you working with a single-file rust extension or a crate? If it's a single-file extension, did you correctly specify the dependency in your header comments?
However, as you can import some parts of druid I'd consider it unlikely that rustimport is the cause of this problem as well - so I'd recommend you to move your issue over to pyo3

**Comment by JoshuaKeegan3** [comment]
So this is actually a problem with rust import which I didn't see coming when you suggested it was py03.

I tried the same setup, with py03 and Maturin build/develop and it gave me an error on the code without the brackets. A reminder that this worked with python and not rust last time. This time, as expected, both the rust code with cargo and the python code with Martin worked with brackets (calling the function).

to simplify the code lets say it is a single file main.rs with the header:

```rust
// rustimport:pyo3
//  Set the library's version and add a dependency:
//: [package]
//: version = "0.1.0"
//:
//: [dependencies]
//: druid = "*"
// --snip--

use druid::{Color, Widget};
use pyo3::prelude::*;
#[pyfunction]
fn test() {
    let window = WindowDesc::new(build_app())
}
fn main(){
  test()
}
```

the the cargo .toml looks like

```toml
[dependencies]
pyo3 = { version = "*", features = ["extension-module"] }
druid = "*"
```

And the loss of deep type information is a py03 feature not a rust import feature and doesn't cause a problem.

**Comment by mityax** [comment]
Hm, that's interesting. I could imagine this is caused by different pyo3 versions, could you try setting the pyo3 version to '*' in your header comments as well?

At the moment the pyo3 rustimport template fixes the pyo3 version to something 0.18.3 if it's unspecified in the header comments. This is to avoid sudden breaking changes taking effect when a new pyo3 version is published.

Curios whether that fixes it. Otherwise, the only thing remaining that might cause this could be some compiler flags that maturin passes to `cargo build`, but I wouldn't know of any that would lead to such behaviour.
